File: okyo/src/main/java/com/commit451/okyo/Okyo.java
package com.commit451.okyo;

import android.support.annotation.NonNull;

import java.io.ByteArrayInputStream;
import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;

import okio.BufferedSink;
import okio.BufferedSource;
import okio.Okio;

/**
 * okio made more ok, yo
 */
public class Okyo {

    public static void writeInputStreamToFile(@NonNull InputStream inputStream, @NonNull File file) throws IOException {
        BufferedSink sink = Okio.buffer(Okio.sink(file));
        BufferedSource source = Okio.buffer(Okio.source(inputStream));
        sink.writeAll(source);
        sink.flush();
        sink.close();
        source.close();
    }

    public static void writeByteArrayToFile(@NonNull byte[] bytes, @NonNull File file) throws IOException{
        BufferedSink sink = Okio.buffer(Okio.sink(file));
        BufferedSource source = Okio.buffer(Okio.source(new ByteArrayInputStream(bytes)));
        sink.writeAll(source);
        sink.flush();
        sink.close();
        source.close();
    }

    public static void writeByteArrayToOutputStream(@NonNull byte[] bytes, @NonNull OutputStream outputStream) throws IOException {
        BufferedSink sink = Okio.buffer(Okio.sink(outputStream));
        BufferedSource source = Okio.buffer(Okio.source(new ByteArrayInputStream(bytes)));
        sink.writeAll(source);
        sink.flush();
        sink.close();
        source.close();
    }

    public static void writeStringToFile(@NonNull String content, @NonNull File file) throws IOException {
        writeByteArrayToFile(content.getBytes(), file);
    }

    public static void writeFileToFile(@NonNull File fileToReadFrom, @NonNull File fileToWriteTo) throws IOException {
        BufferedSink sink = Okio.buffer(Okio.sink(fileToWriteTo));
        BufferedSource source = Okio.buffer(Okio.source(fileToReadFrom));
        sink.writeAll(source);
        sink.flush();
        sink.close();
        source.close();
    }

    @NonNull
    public static String readInputStreamAsString(@NonNull InputStream inputStream) throws IOException {
        BufferedSource source = Okio.buffer(Okio.source(inputStream));
        String content =  source.readUtf8();
        source.close();
        return content;
    }

    @NonNull
    public static String readFileAsString(@NonNull File file) throws IOException {
        BufferedSource source = Okio.buffer(Okio.source(file));
        String content =  source.readUtf8();
        source.close();
        return content;
    }

    @NonNull
    public static byte[] readFileAsByteArray(@NonNull File file) throws IOException {
        BufferedSource source = Okio.buffer(Okio.source(file));
        byte[] bytes = source.readByteArray();
        source.close();
        return bytes;
    }
}

File: src/main/kotlin/com/commit451/okyo/Okyo.kt
@file:Suppress("unused")

package com.commit451.okyo

import okio.buffer
import okio.sink
import okio.source
import java.io.*

@Suppress("NOTHING_TO_INLINE")
/**
 * okio made more ok, yo
 */
object Okyo {

    @JvmStatic
    @Throws(IOException::class)
    inline fun writeInputStreamToFile(inputStream: InputStream, file: File) {
        val sink = file.sink().buffer()
        val source = inputStream.source().buffer()
        sink.writeAll(source)
        sink.flush()
        sink.close()
        source.close()
    }

    @JvmStatic
    @Throws(IOException::class)
    inline fun writeByteArrayToFile(bytes: ByteArray, file: File) {
        val sink = file.sink().buffer()
        val source = ByteArrayInputStream(bytes).source().buffer()
        sink.writeAll(source)
        sink.flush()
        sink.close()
        source.close()
    }

    @JvmStatic
    @Throws(IOException::class)
    inline fun writeStringToOutputStream(string: String, outputStream: OutputStream) {
        return writeByteArrayToOutputStream(string.toByteArray(), outputStream)
    }

    @JvmStatic
    @Throws(IOException::class)
    inline fun writeByteArrayToOutputStream(bytes: ByteArray, outputStream: OutputStream) {
        val sink = outputStream.sink().buffer()
        val source = ByteArrayInputStream(bytes).source().buffer()
        sink.writeAll(source)
        sink.flush()
        sink.close()
        source.close()
    }

    @JvmStatic
    @Throws(IOException::class)
    inline fun writeStringToFile(content: String, file: File) {
        writeByteArrayToFile(content.toByteArray(), file)
    }

    @JvmStatic
    @Throws(IOException::class)
    inline fun writeFileToFile(fileToReadFrom: File, fileToWriteTo: File) {
        val sink = fileToWriteTo.sink().buffer()
        val source = fileToReadFrom.source().buffer()
        sink.writeAll(source)
        sink.flush()
        sink.close()
        source.close()
    }

    @JvmStatic
    @Throws(IOException::class)
    inline fun writeFileToOutputStream(fileToReadFrom: File, outputStream: OutputStream) {
        val sink = outputStream.sink().buffer()
        val source = fileToReadFrom.source().buffer()
        sink.writeAll(source)
        sink.flush()
        sink.close()
        source.close()
    }

    @JvmStatic
    @Throws(IOException::class)
    inline fun readInputStreamAsString(inputStream: InputStream): String {
        val source = inputStream.source().buffer()
        val content = source.readUtf8()
        source.close()
        return content
    }

    @JvmStatic
    @Throws(IOException::class)
    inline fun readFileAsString(file: File): String {
        val source = file.source().buffer()
        val content = source.readUtf8()
        source.close()
        return content
    }

    @JvmStatic
    @Throws(IOException::class)
    inline fun readFileAsByteArray(file: File): ByteArray {
        val source = file.source().buffer()
        val bytes = source.readByteArray()
        source.close()
        return bytes
    }
}
